lib/auth/join_bound_keypair.go

```diff
@@ -892,16 +888,21 @@ func (a *Server) RegisterUsingBoundKeypairMethod(
 			return nil, trace.AccessDenied("join state verification failed")
 		}
 
+		// Join state verification will check the instance IDs in the token and
+		// join state document, but as a sanity check, we'll also ensure it
+		// matches the value extracted from the certs.
+		//
+		// It should not be possible for this check to fail at this point, as
+		// any event that might have cycled bot instance IDs should have also
+		// modified the join state causing a failure above. In any case, we'll
+		// keep this as a sanity check.
+		if status.BoundBotInstanceID != req.JoinRequest.BotInstanceID {
+			return nil, trace.AccessDenied("bot instance mismatch")
+		}
+
 		// Nothing else to do, no key change, no additional audit event; regular
 		// bot join event will be emitted later.
 	case hasBoundPublicKey && hasBoundBotInstance && !hasIncomingBotInstance:
-		// Hard rejoin case, the client identity expired and a new bot instance
-		// is required. Consumes a rejoin.
-		if recoveryMode == boundkeypair.RecoveryModeStandard && !hasJoinsRemaining {
-			// Recovery limit only applies in "standard" mode.
-			return nil, trace.AccessDenied("no rejoins remaining")
-		}
```

Comment:
This isn't strictly related to this fix, but is a similar enough issue that I figured I'd also fix it here - no need to reveal information about the token until we've at least somewhat verified the client's identity.